# Chapter 05. 코사인 유사도로 비슷한 책 추천하기

Chapter 04에서는 TF-IDF 벡터와 Naive Bayes로 도서 분야를 분류했다.
이번 Chapter에서는 **같은 TF-IDF 표현을 추천에 활용**한다.

```
Chapter 04   도서 제목 -> TF-IDF -> 분야 분류
Chapter 05   도서 제목 -> TF-IDF -> Cosine Similarity -> 비슷한 도서 추천
```

이번 Chapter의 목표는 복잡한 추천 알고리즘을 배우는 것이 아니다.
도서 제목을 벡터로 표현하고, 선택한 도서와 다른 도서의 유사도를 계산해
**Top 5를 추천하는 전체 흐름**을 이해한다.

## 최종 결과물

```
chapter05.ipynb
chapter05_recommendations.csv
```

---
# 1. 학습 목표

- 콘텐츠 기반 추천이 무엇인지 설명할 수 있다.
- 도서 제목을 TF-IDF 벡터로 변환할 수 있다.
- 코사인 유사도의 의미를 설명할 수 있다.
- 선택한 도서와 전체 도서의 유사도를 계산할 수 있다.
- 자기 자신을 추천 결과에서 제외할 수 있다.
- 유사도 순으로 정렬하여 Top 5를 만들 수 있다.
- 추천 로직을 함수로 만들 수 있다.
- 제목 기반 추천의 한계를 설명할 수 있다.
- Chapter 06의 Streamlit 앱과 연결할 수 있다.

## 전체 흐름

```
데이터 준비 -> 상품명 TF-IDF -> 기준 도서 선택 -> 전체 도서와 Cosine Similarity
-> 유사도 내림차순 정렬 -> 자기 자신 제외 -> Top 5 -> 추천 함수 -> Streamlit 연결
```

---
# 실습 1. 라이브러리 준비하기

## 해야 할 일

추천에 필요한 numpy, pandas, TF-IDF, 코사인 유사도 기능을 불러온다.
필요한 패키지가 없다면 가상환경에서 설치한다.

```
python -m pip install pandas numpy scikit-learn
```

## 프롬프트 예시

```
도서 제목으로 비슷한 책을 추천하려고 합니다.
numpy, pandas, TfidfVectorizer, 그리고 sklearn의 cosine_similarity를 불러오는
import 코드를 작성해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [1]:
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

print("라이브러리 준비 완료")

라이브러리 준비 완료


### 작업 폴더를 맞춘다

이 저장소의 `.vscode/settings.json`에 `"jupyter.notebookFileRoot": "${workspaceFolder}"`가 있어서,
Notebook을 실행하면 작업 폴더가 노트북이 있는 곳이 아니라 **저장소 루트**가 된다.
그래서 파일 이름만 적으면 `FileNotFoundError`가 난다.
아래 셀에서 작업 폴더를 `book-text-ml`로 맞춘 뒤 진행한다.

## 프롬프트 예시

```
VS Code에서 Jupyter Notebook을 쓰고 있습니다.
설정 때문에 작업 폴더가 노트북이 있는 곳이 아니라 저장소 루트가 됩니다.
노트북이 들어 있는 book-text-ml 폴더를 찾아 그쪽으로 이동하고 싶습니다.
폴더가 저장소 루트에 있을 수도 있고 notebooks 안에 있을 수도 있습니다.
마지막에 현재 작업 폴더와 데이터 파일 존재 여부를 출력해 주세요.
try/except 예외 처리는 넣지 마세요.
초보자가 읽기 쉽게 10줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [2]:
import os
from pathlib import Path

# 노트북이 있는 book-text-ml 폴더를 찾아 그쪽으로 이동한다.
# 폴더가 저장소 루트에 있든 notebooks 안에 있든 모두 찾는다.
if Path.cwd().name != "book-text-ml":
    for candidate in [Path("book-text-ml"), *Path(".").glob("*/book-text-ml")]:
        if candidate.is_dir():
            os.chdir(candidate)
            break

print("작업 폴더:", Path.cwd())
print("데이터 파일 있음:", Path("book_bestseller_clean.csv").exists())

작업 폴더: C:\dev\llm-data-analysis-course\notebooks\book-text-ml
데이터 파일 있음: True


---
# 실습 2. 데이터 불러오기

## 해야 할 일

Chapter 01에서 만든 전처리 파일을 사용한다.
이번 추천에서 가장 중요한 컬럼은 **상품명**이다.

추천 결과를 보기 좋게 하기 위해 데이터에 존재한다면 `인물`, `출판사`, `분야`도 함께 사용한다.

## 프롬프트 예시

```
Python과 pandas를 처음 배우고 있습니다.
현재 Notebook과 같은 폴더에 book_bestseller_clean.csv 파일이 있습니다.
utf-8-sig 인코딩으로 df_books라는 DataFrame으로 불러오고,
데이터 크기, 컬럼 목록, 상품명 결측치 개수, 상품명 앞 10개를 확인하고 싶습니다.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [3]:
DATA_PATH = "book_bestseller_clean.csv"
df_books = pd.read_csv(
    DATA_PATH,
    encoding="utf-8-sig",
)

print("데이터 크기:", df_books.shape)
print("컬럼:", df_books.columns.tolist())
print("상품명 결측치:", df_books["상품명"].isna().sum())
df_books[["상품명"]].head(10)

데이터 크기: (986, 8)
컬럼: ['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야']
상품명 결측치: 0


,상품명
0,"세네카, 오늘을 빼앗기고 있는 당신에게"
1,흔한남매 23
2,머니 트렌드 2027
3,싯다르타
4,한국사 이상현상 연구원(일반판)
5,테오
6,쇼펜하우어 인생수업(30만 부 기념 개정증보판)
7,니체의 초월자
8,판매의 법칙
9,마음의 어휘력


---
# 실습 3. 추천용 데이터 준비하기

## 해야 할 일

상품명을 정리하고 비어 있는 제목을 뺀 뒤 **index를 다시 매긴다.**

여기서 `reset_index(drop=True)`가 중요하다.

```
DataFrame index 0  <->  TF-IDF matrix row 0
DataFrame index 1  <->  TF-IDF matrix row 1
```

DataFrame의 index와 TF-IDF 행 번호가 **같은 도서를 가리켜야** 한다.

## 프롬프트 예시

```
DataFrame df_books를 복사해서 추천용 df_reco를 만들고 싶습니다.
상품명의 결측치는 빈 문자열로 바꾸고, 문자열로 통일하고, 앞뒤 공백을 제거한 뒤
빈 제목은 빼고 index를 0부터 다시 매겨 주세요.
TF-IDF 행 번호와 DataFrame index가 같은 책을 가리켜야 합니다.
마지막에 추천에 사용할 도서 수를 출력해 주세요.
try/except 예외 처리는 넣지 마세요.
초보자가 읽기 쉽게 10줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [4]:
df_reco = df_books.copy()
df_reco["상품명"] = (
    df_reco["상품명"]
    .fillna("")
    .astype(str)
    .str.strip()
)
df_reco = (
    df_reco[df_reco["상품명"] != ""]
    .reset_index(drop=True)
)

print("추천에 사용할 도서 수:", len(df_reco))

추천에 사용할 도서 수: 986


---
# 실습 4. 콘텐츠 기반 추천 이해하기

이번 실습에서는 **콘텐츠 기반 추천(Content-Based Recommendation)** 을 사용한다.

```
선택한 항목의 특징
        ↓
비슷한 특징을 가진 항목 찾기
```

이번에는 도서의 특징으로 **상품명 텍스트**를 사용한다.

```
도서 A 제목 -> TF-IDF 벡터 A
도서 B 제목 -> TF-IDF 벡터 B
A와 B 벡터가 비슷한가? -> 비슷하면 추천 후보
```

이번 추천은 다음 정보를 **사용하지 않는다.**

```
사용자 구매 이력 / 클릭 이력 / 평점 / 판매량 / 독자 취향
```

따라서 개인화 추천이라기보다 **도서 제목 기반 유사 도서 추천**이라고 표현하는 것이 정확하다.

---
# 실습 5. 코사인 유사도 이해하기

## 해야 할 일

코사인 유사도는 **두 벡터의 방향이 얼마나 비슷한지** 비교한다.

| 값 | 의미 |
|---|---|
| 1에 가까움 | 현재 벡터 표현에서 매우 비슷함 |
| 0에 가까움 | 공통 특징이 적음 |

간단한 벡터로 먼저 확인한다.

## 프롬프트 예시

```
코사인 유사도의 의미를 작은 예제로 확인하고 싶습니다.
[1, 1], [2, 2], [1, 0] 세 벡터를 numpy 배열로 만들고
cosine_similarity로 서로의 유사도를 표로 보여 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [5]:
vectors = np.array([
    [1, 1],
    [2, 2],
    [1, 0],
])

cosine_similarity(vectors)

array([[1.        , 1.        , 0.70710678],
       [1.        , 1.        , 0.70710678],
       [0.70710678, 0.70710678, 1.        ]])

### 결과 읽기

- `[1, 1]`과 `[2, 2]`는 크기는 다르지만 **같은 방향**이라 유사도가 1이다.
- `[1, 1]`과 `[1, 0]`은 45도 차이가 나서 약 0.707이다.
- 대각선(자기 자신과의 비교)은 모두 1이다.

**자기 자신과 비교한 유사도는 일반적으로 1**이다.
이 때문에 추천에서는 나중에 자기 자신을 반드시 제외한다.

---
# 실습 6. 실제 도서 제목을 TF-IDF로 변환하기

## 해야 할 일

```
행 = 도서 / 열 = 단어 / 값 = TF-IDF 가중치
```

Chapter 04에서는 분류 성능을 **평가**하기 때문에 train/test를 먼저 나누고 train에만 fit했다.
이번 Chapter는 **현재 카탈로그 안에서 서로 비슷한 도서를 찾는** 실습이므로
전체 추천 대상 제목을 이용해 TF-IDF 표현을 만든다.

## 프롬프트 예시

```
추천용 DataFrame df_reco의 상품명으로 TF-IDF 행렬을 만들고 싶습니다.
titles 변수에 상품명을 담고, TfidfVectorizer로 tfidf_matrix를 만든 뒤
도서 수와 단어 수를 출력해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [6]:
titles = df_reco["상품명"]

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(titles)

print("도서 수:", tfidf_matrix.shape[0])
print("단어 수:", tfidf_matrix.shape[1])

도서 수: 986
단어 수: 2179


---
# 실습 7. 기준 도서 선택하기

## 해야 할 일

먼저 한 권을 index로 선택한다.
**index 번호만 보지 말고 실제 제목을 반드시 확인한다.**

## 프롬프트 예시

```
추천의 기준이 될 도서를 index 번호로 하나 고르고 싶습니다.
0번 도서를 선택해서 그 상품명을 selected_title에 담고 출력해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [7]:
selected_index = 0
selected_title = df_reco.loc[
    selected_index,
    "상품명",
]

print("선택 도서:", selected_title)

선택 도서: 세네카, 오늘을 빼앗기고 있는 당신에게


---
# 실습 8. 선택 도서와 전체 도서의 유사도 계산하기

## 해야 할 일

선택한 도서의 벡터를 가져오고, 전체 도서와 코사인 유사도를 계산한다.

```
similarity_scores[0]  <->  df_reco.iloc[0]
similarity_scores[1]  <->  df_reco.iloc[1]
```

각 유사도 점수와 원본 도서가 **정확하게 대응**해야 한다.

## 프롬프트 예시

```
TF-IDF 행렬 tfidf_matrix와 선택한 도서 번호 selected_index가 있습니다.
선택한 도서의 벡터만 selected_vector로 가져오고 싶습니다.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [8]:
selected_vector = tfidf_matrix[selected_index]

## 프롬프트 예시

```
선택한 도서의 벡터 selected_vector와 전체 tfidf_matrix가 있습니다.
선택 도서와 전체 도서의 코사인 유사도를 계산해서 1차원 배열 similarity_scores로 만들고,
유사도 개수와 전체 도서 수를 출력해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [9]:
similarity_scores = cosine_similarity(
    selected_vector,
    tfidf_matrix,
).flatten()

print("유사도 개수:", len(similarity_scores))
print("전체 도서 수:", len(df_reco))

유사도 개수: 986
전체 도서 수: 986


## 프롬프트 예시

```
유사도 배열 similarity_scores와 DataFrame df_reco가 있습니다.
두 길이가 같은지, 앞의 3개 점수가 어떤 도서와 대응하는지,
그리고 선택한 도서와 자기 자신의 유사도가 1.0인지 확인하고 싶습니다.
try/except 예외 처리는 넣지 마세요.
초보자가 읽기 쉽게 10줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [10]:
# 두 값이 같은지, 점수와 도서가 순서대로 대응하는지 직접 확인한다
print("두 값이 같은가?:", len(similarity_scores) == len(df_reco))
print()
for i in range(3):
    print(f"similarity_scores[{i}] = {similarity_scores[i]:.4f}  <->  {df_reco.iloc[i]['상품명']}")
print()
print("자기 자신과의 유사도:", round(float(similarity_scores[selected_index]), 4), "(1.0이어야 한다)")

두 값이 같은가?: True

similarity_scores[0] = 1.0000  <->  세네카, 오늘을 빼앗기고 있는 당신에게
similarity_scores[1] = 0.0000  <->  흔한남매 23
similarity_scores[2] = 0.0000  <->  머니 트렌드 2027

자기 자신과의 유사도: 1.0 (1.0이어야 한다)


---
# 실습 9. 유사도가 높은 순서 확인하기

## 해야 할 일

먼저 결과를 제목과 함께 확인한다.

가장 위에는 일반적으로 **선택한 도서 자기 자신**이 나타난다(similarity ≈ 1.0).
이것은 오류가 아니라 **정상적인 결과**다.

## 프롬프트 예시

```
유사도 배열 similarity_scores와 DataFrame df_reco가 있습니다.
index, 상품명, similarity 세 컬럼을 가진 표를 만들고
유사도가 높은 순서로 정렬해서 상위 10개를 보여 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [11]:
score_df = pd.DataFrame({
    "index": np.arange(len(df_reco)),
    "상품명": df_reco["상품명"],
    "similarity": similarity_scores,
})

score_df.sort_values(
    "similarity",
    ascending=False,
).head(10)

,index,상품명,similarity
0,0,"세네카, 오늘을 빼앗기고 있는 당신에게",1.000000
587,587,"오늘을 빼앗기는 당신에게, 세네카",0.668201
554,554,돌이킬 수 있는,0.252116
659,659,남아 있는 나날,0.203866
521,521,비전공자도 이해할 수 있는 LLM 수업,0.161707
349,349,품격 있는 대화를 위한 지식 브리핑,0.147451
928,928,비전공자도 이해할 수 있는 AI 지식(10만부 기념 개정판),0.132410
4,4,한국사 이상현상 연구원(일반판),0.000000
5,5,테오,0.000000
6,6,쇼펜하우어 인생수업(30만 부 기념 개정증보판),0.000000


---
# 실습 10. 자기 자신을 제외하고 Top 5 만들기

## 해야 할 일

유사도가 높은 순서의 index를 구하고, 자기 자신을 제외한 뒤 상위 5개만 선택한다.

```
유사도 계산 -> 높은 순서로 정렬 -> 자기 자신 제외 -> Top 5
```

**추천의 핵심 로직은 이것으로 완성된다.**

## 프롬프트 예시

```
유사도 배열 similarity_scores가 있습니다.
유사도가 높은 순서대로 index를 정렬해서 sorted_indices에 담고 싶습니다.
argsort는 오름차순이니 순서를 뒤집어 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [12]:
sorted_indices = similarity_scores.argsort()[::-1]

## 프롬프트 예시

```
유사도 높은 순으로 정렬된 index 목록 sorted_indices와
선택한 도서 번호 selected_index가 있습니다.
자기 자신은 빼고 상위 5개 index만 recommended_indices로 골라 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [13]:
recommended_indices = [
    idx
    for idx in sorted_indices
    if idx != selected_index
][:5]

recommended_indices

[np.int64(587), np.int64(554), np.int64(659), np.int64(521), np.int64(349)]

## 프롬프트 예시

```
추천할 index 목록 recommended_indices와 유사도 배열 similarity_scores가 있습니다.
df_reco에서 추천 도서의 상품명만 뽑고
similarity 컬럼에 소수점 4자리로 유사도를 붙인 표를 만들어 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [14]:
result = df_reco.loc[
    recommended_indices,
    ["상품명"],
].copy()

result["similarity"] = [
    round(float(similarity_scores[idx]), 4)
    for idx in recommended_indices
]

result

,상품명,similarity
587,"오늘을 빼앗기는 당신에게, 세네카",0.6682
554,돌이킬 수 있는,0.2521
659,남아 있는 나날,0.2039
521,비전공자도 이해할 수 있는 LLM 수업,0.1617
349,품격 있는 대화를 위한 지식 브리핑,0.1475


---
# 실습 11. 추천 결과에 메타데이터 추가하기

## 해야 할 일

제목만 보는 것보다 저자, 출판사, 분야를 함께 보면 추천 결과를 검토하기 쉽다.

`available_columns`는 **데이터에 실제로 있는 컬럼만** 골라 쓴다.
이 데이터는 Chapter 01에서 `인물`을 `저자`로 바꿔 저장했기 때문에
`인물`은 자동으로 빠지고 `상품명`, `출판사`, `분야`가 사용된다.

## 프롬프트 예시

```
추천 결과에 저자, 출판사, 분야도 함께 보여주고 싶습니다.
상품명, 인물, 출판사, 분야 중에서 df_reco에 실제로 있는 컬럼만 골라
available_columns로 만들고, 그 컬럼들로 추천 결과 표를 만든 뒤
similarity 컬럼을 소수점 4자리로 붙여 주세요.
try/except 예외 처리는 넣지 마세요.
초보자가 읽기 쉽게 10줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [15]:
available_columns = [
    column
    for column in [
        "상품명",
        "인물",
        "출판사",
        "분야",
    ]
    if column in df_reco.columns
]

result = df_reco.loc[
    recommended_indices,
    available_columns,
].copy()

result["similarity"] = [
    round(float(similarity_scores[idx]), 4)
    for idx in recommended_indices
]

result

,상품명,출판사,분야,similarity
587,"오늘을 빼앗기는 당신에게, 세네카",유페이퍼,미분류,0.6682
554,돌이킬 수 있는,아작,소설,0.2521
659,남아 있는 나날,민음사,소설,0.2039
521,비전공자도 이해할 수 있는 LLM 수업,비즈니스북스,과학,0.1617
349,품격 있는 대화를 위한 지식 브리핑,북플레저,인문,0.1475


## 프롬프트 예시

```
선택한 도서의 제목과 추천 결과 표 result를 위아래로 나란히 보고 싶습니다.
표 형태가 유지되도록 display()를 사용해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [16]:
print("선택 도서:")
print(selected_title)
print("\n추천 도서:")
display(result)

선택 도서:
세네카, 오늘을 빼앗기고 있는 당신에게

추천 도서:


,상품명,출판사,분야,similarity
587,"오늘을 빼앗기는 당신에게, 세네카",유페이퍼,미분류,0.6682
554,돌이킬 수 있는,아작,소설,0.2521
659,남아 있는 나날,민음사,소설,0.2039
521,비전공자도 이해할 수 있는 LLM 수업,비즈니스북스,과학,0.1617
349,품격 있는 대화를 위한 지식 브리핑,북플레저,인문,0.1475


### 직접 확인할 것

- 공통 단어가 있는가?
- 같은 시리즈인가?
- 분야도 비슷한가?
- 단순히 제목 표현만 비슷한 것은 아닌가?

### 확인 결과

`세네카, 오늘을 빼앗기고 있는 당신에게`를 골랐을 때 1위는
`오늘을 빼앗기는 당신에게, 세네카`(유사도 0.6682)였다.
출판사만 다른 **사실상 같은 책**이다.

2위부터는 유사도가 0.25 이하로 뚝 떨어진다.
`돌이킬 수 있는`, `남아 있는 나날`은 `있는`이라는 **흔한 단어 하나만 겹쳐서** 추천된 것이다.
세네카와 주제가 비슷해서가 아니다.

---
# 실습 12. 추천 함수 만들기

## 해야 할 일

지금까지의 코드를 하나의 함수로 묶는다. 이 함수가 하는 일은 다섯 단계다.

```
선택한 도서 받기 -> 전체 도서와 유사도 계산 -> 높은 순서 정렬
-> 자기 자신과 동일 제목 제외 -> Top N 반환
```

실습 10과 달리 **동일한 제목도 제외**한다.
같은 책이 출판사나 분야만 다르게 여러 번 등록된 경우가 있기 때문이다.

## 프롬프트 예시

```
유사도 배열 similarity_scores가 있습니다.
유사도가 높은 순서대로 index를 정렬해서 sorted_indices에 담고 싶습니다.
argsort는 오름차순이니 순서를 뒤집어 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [17]:
def recommend_books(
    selected_index,
    df,
    tfidf_matrix,
    top_n=5,
):
    if selected_index not in df.index:
        raise IndexError(
            f"유효하지 않은 index입니다: {selected_index}"
        )

    selected_title = df.loc[
        selected_index,
        "상품명",
    ]

    similarity_scores = cosine_similarity(
        tfidf_matrix[selected_index],
        tfidf_matrix,
    ).flatten()

    sorted_indices = similarity_scores.argsort()[::-1]

    recommended_indices = []

    for idx in sorted_indices:
        # 선택한 자기 자신 제외
        if idx == selected_index:
            continue

        # 동일한 제목도 제외
        if df.loc[idx, "상품명"] == selected_title:
            continue

        recommended_indices.append(idx)

        if len(recommended_indices) >= top_n:
            break

    columns = [
        column
        for column in [
            "상품명",
            "인물",
            "출판사",
            "분야",
        ]
        if column in df.columns
    ]

    result = df.loc[
        recommended_indices,
        columns,
    ].copy()

    result["similarity"] = [
        round(float(similarity_scores[idx]), 4)
        for idx in recommended_indices
    ]

    return result.reset_index(drop=True)

print("recommend_books 함수 준비 완료")

recommend_books 함수 준비 완료


## 프롬프트 예시

```
df_reco에 같은 제목이 두 번 이상 등록된 도서가 있는지 확인하고 싶습니다.
중복된 제목을 가진 행을 모두 골라 제목 순으로 정렬하고,
그 개수와 상품명·출판사·분야 앞 6개를 보여 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [18]:
# 동일 제목 제외가 왜 필요한지 확인한다
# 같은 제목이 출판사/분야만 다르게 두 번 등록된 책이 있다
dup_titles = df_reco[df_reco["상품명"].duplicated(keep=False)].sort_values("상품명")

print("같은 제목이 2번 이상 등록된 도서 수:", len(dup_titles))
dup_titles[["상품명", "출판사", "분야"]].head(6)

같은 제목이 2번 이상 등록된 도서 수: 88


,상품명,출판사,분야
714,2026 하반기 해커스 GSAT 삼성직무적성검사 실전모의고사 12회분,해커스잡,취업/수험서
682,2026 하반기 해커스 GSAT 삼성직무적성검사 실전모의고사 12회분,해커스잡,미분류
102,2026 해커스 GSAT 삼성직무적성검사 통합 기본서 최신기출유형+실전모의고사 (수...,해커스잡,취업/수험서
138,2026 해커스 GSAT 삼성직무적성검사 통합 기본서 최신기출유형+실전모의고사 (수...,해커스잡,미분류
377,"AI, 신의 탄생 인간의 종말",상상스퀘어,미분류
99,"AI, 신의 탄생 인간의 종말",상상스퀘어,경제/경영


### 같은 책이 두 번 들어 있다

같은 제목·같은 출판사인데 **한쪽은 실제 분야, 한쪽은 `미분류`** 로 등록된 경우가 많다.
Chapter 01에서 분야가 비어 있던 책을 `미분류`로 채웠는데,
그중 상당수가 **이미 다른 행에 분야와 함께 있는 같은 책**이었던 것이다.

실습 10의 단순한 방식(자기 자신만 제외)으로는 이 복제본이 유사도 1.0으로 1위에 올라온다.
그래서 추천 함수에서는 **동일한 제목도 제외**한다.

---
# 실습 13. 추천 함수 실행하기

## 해야 할 일

함수를 실행하고 다음 항목을 확인한다.

- 자기 자신이 제외되었는가?
- 추천 결과가 최대 5개인가?
- similarity가 높은 순서인가?
- 실제 원본 도서 정보와 일치하는가?

## 프롬프트 예시

```
recommend_books 함수와 df_reco, tfidf_matrix가 있습니다.
0번 도서를 선택해서 제목을 출력하고,
함수로 Top 5 추천 결과를 만들어 recommendations로 보여 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [19]:
selected_index = 0
print(
    "선택 도서:",
    df_reco.loc[selected_index, "상품명"],
)

recommendations = recommend_books(
    selected_index=selected_index,
    df=df_reco,
    tfidf_matrix=tfidf_matrix,
    top_n=5,
)

recommendations

선택 도서: 세네카, 오늘을 빼앗기고 있는 당신에게


,상품명,출판사,분야,similarity
0,"오늘을 빼앗기는 당신에게, 세네카",유페이퍼,미분류,0.6682
1,돌이킬 수 있는,아작,소설,0.2521
2,남아 있는 나날,민음사,소설,0.2039
3,비전공자도 이해할 수 있는 LLM 수업,비즈니스북스,과학,0.1617
4,품격 있는 대화를 위한 지식 브리핑,북플레저,인문,0.1475


## 프롬프트 예시

```
추천 결과 recommendations가 제대로 나왔는지 점검하고 싶습니다.
df_reco 행 수와 tfidf_matrix 행 수가 같은지 assert로 확인하고,
추천 개수, 자기 자신이 제외됐는지, 유사도가 내림차순인지 출력해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [20]:
assert df_reco.shape[0] == tfidf_matrix.shape[0]

print("행 수 일치 확인 통과")
print("추천 개수:", len(recommendations))
print("자기 자신 제외됐는가?:", selected_title not in recommendations["상품명"].tolist())
print("유사도 내림차순인가?:", recommendations["similarity"].is_monotonic_decreasing)

행 수 일치 확인 통과
추천 개수: 5
자기 자신 제외됐는가?: True
유사도 내림차순인가?: True


---
# 실습 14. 추천 결과 저장하기

저장한 뒤 **다시 읽어** 정상적으로 저장되었는지 확인한다.

## 프롬프트 예시

```
추천 결과 recommendations를 chapter05_recommendations.csv로 저장하고 싶습니다.
utf-8-sig 인코딩을 쓰고 인덱스는 저장하지 말아 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [21]:
recommendations.to_csv(
    "chapter05_recommendations.csv",
    index=False,
    encoding="utf-8-sig",
)

print("저장 완료")

저장 완료


## 프롬프트 예시

```
방금 저장한 chapter05_recommendations.csv를 다시 읽어서
제대로 저장됐는지 확인하고 싶습니다.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [22]:
pd.read_csv(
    "chapter05_recommendations.csv",
    encoding="utf-8-sig",
)

,상품명,출판사,분야,similarity
0,"오늘을 빼앗기는 당신에게, 세네카",유페이퍼,미분류,0.6682
1,돌이킬 수 있는,아작,소설,0.2521
2,남아 있는 나날,민음사,소설,0.2039
3,비전공자도 이해할 수 있는 LLM 수업,비즈니스북스,과학,0.1617
4,품격 있는 대화를 위한 지식 브리핑,북플레저,인문,0.1475


---
# 실습 15. 추천 결과의 한계 이해하기

현재 추천은 **상품명 텍스트만** 사용한다. 따라서 다음과 같은 한계가 있다.

```
같은 분야라도 제목 단어가 다르면       -> 유사도가 낮을 수 있음
다른 분야라도 공통 표현이 많으면       -> 유사도가 높을 수 있음
```

또한 코사인 유사도는 **취향 점수가 아니다.**

```
유사도 높음 ≠ 사용자가 반드시 좋아함
           ≠ 더 좋은 책
           ≠ 구매 가능성이 높음
```

| 올바른 표현 | 피해야 할 표현 |
|---|---|
| 현재 TF-IDF 제목 표현 기준으로 유사도가 높은 도서입니다. | 사용자가 가장 좋아할 책입니다. |

이 한계가 실제 데이터에서 얼마나 나타나는지는 아래 보충 분석에서 숫자로 확인한다.

---
# 보충 1. 여러 도서로 추천을 돌려보기

## 해야 할 일

기준 도서 한 권(세네카)만 보고 추천이 잘 된다고 판단할 수 없다.
성격이 다른 도서 몇 권을 골라 추천 결과를 비교한다.

## 프롬프트 예시

```
recommend_books 함수와 df_reco, tfidf_matrix가 있습니다.
1, 2, 3, 5, 8번 도서를 차례로 기준 도서로 골라
각각의 제목·분야와 Top 5 추천 결과(유사도, 제목, 분야)를 출력해 비교하고 싶습니다.
try/except 예외 처리는 넣지 마세요.
초보자가 읽기 쉽게 10줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [23]:
for idx in [1, 2, 3, 5, 8]:
    print(f"[{idx}] {df_reco.loc[idx, '상품명']}  ({df_reco.loc[idx, '분야']})")
    recs = recommend_books(idx, df_reco, tfidf_matrix, top_n=5)
    for _, row in recs.iterrows():
        print(f"      {row['similarity']:.3f}  {row['상품명'][:38]}  ({row['분야']})")
    print()

[1] 흔한남매 23  (어린이(초등))
      0.408  흔한남매 22  (어린이(초등))
      0.395  흔한남매 이무기 4  (어린이(초등))
      0.340  흔한남매 세계사 탐험대 7  (어린이(초등))
      0.332  흔한남매 방방곡곡 한국사 1  (어린이(초등))
      0.228  흔한남매 과학 탐험대 17: 뇌와 호르몬  (어린이(초등))

[2] 머니 트렌드 2027  (경제/경영)
      0.539  부동산 트렌드 2027  (경제/경영)
      0.524  트렌드 코리아 2027  (경제/경영)
      0.131  2027 Trend Preview  (강연)
      0.130  2027 공단기 심슨 독해  (취업/수험서)
      0.129  2027 공단기 심슨 보카  (취업/수험서)

[3] 싯다르타  (소설)
      0.363  싯다르타(초판본)(1922년 오리지널 초판본 표지디자인)  (소설)
      0.000  교보문고 북커버 키트(브라운)  (교보문고 굿즈)
      0.000  문해내공  (미분류)
      0.000  이효석의 21세기 투자법  (미분류)
      0.000  [1+1] 볼륨앤소프트 3겹데코 천연펄프 화장지 30롤  (홈/라이프)

[5] 테오  (소설)
      0.000  세종의 나라 2  (미분류)
      0.000  교보문고 북커버 키트(브라운)  (교보문고 굿즈)
      0.000  문해내공  (미분류)
      0.000  이효석의 21세기 투자법  (미분류)
      0.000  [1+1] 볼륨앤소프트 3겹데코 천연펄프 화장지 30롤  (홈/라이프)

[8] 판매의 법칙  (경제/경영)
      0.430  불변의 법칙  (경제/경영)
      0.406  침투율의 법칙  (경제/경영)
      0.303  인간 본성의 법칙 (리커버:K)  (인문)
      0.275  부자 아빠 투자 불변의 법칙  (미분류)
     

### 결과에서 보이는 네 가지 유형

**1. 시리즈는 잘 찾는다** — `흔한남매 23`
```
0.408 흔한남매 22 / 0.395 흔한남매 이무기 4 / 0.340 흔한남매 세계사 탐험대 7 ...
```
추천 5권이 모두 흔한남매 시리즈였다. 제목에 **고유한 시리즈명**이 있으면 제목 기반 추천이 잘 작동한다.

**2. 주제어가 겹치면 그럴듯하다** — `판매의 법칙`
```
0.430 불변의 법칙 / 0.406 침투율의 법칙 / 0.303 인간 본성의 법칙 ...
```
`법칙`이라는 단어로 경제/경영·인문 도서가 묶였다.

**3. 연도 숫자에 끌려간다** — `머니 트렌드 2027`
```
0.539 부동산 트렌드 2027 / 0.524 트렌드 코리아 2027 / 0.130 2027 공단기 심슨 독해 ...
```
위의 두 권은 좋은 추천이지만, 3위부터는 **`2027`만 겹치는 수험서**가 나왔다.
Chapter 03에서 본 것처럼 기본 TF-IDF가 연도를 단어로 취급하기 때문이다.

**4. 겹치는 단어가 없으면 의미 없는 추천이 나온다** — `테오`
```
0.000 세종의 나라 2 / 0.000 교보문고 북커버 키트 / 0.000 문해내공 / 0.000 ... 화장지 30롤
```
유사도가 전부 0인데도 5권을 채워서 **화장지와 북커버 키트**를 추천했다.
유사도가 모두 같으면(0) 정렬 순서가 의미 없어지는데, 함수는 그래도 5개를 채우기 때문이다.

---
# 보충 2. 전체 도서의 추천 품질 점검하기

## 해야 할 일

몇 권만 보면 우연일 수 있다. **986권 모두**에 대해 추천을 돌려서 다음을 확인한다.

- 1위 추천의 유사도가 얼마나 되는가
- 1위 추천조차 유사도가 0인 책(겹치는 단어가 전혀 없는 책)은 몇 권인가
- 추천 5권이 기준 도서와 **같은 분야**인 비율은 얼마인가 (추천이 주제를 잘 따라가는지)

## 프롬프트 예시

```
TF-IDF 행렬 tfidf_matrix가 있습니다.
모든 도서끼리의 코사인 유사도를 한 번에 계산해 all_scores로 만들고
그 크기를 출력해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [24]:
# 986권 전체의 유사도를 한 번에 계산한다 (986 x 986)
all_scores = cosine_similarity(tfidf_matrix)
print("전체 유사도 행렬 크기:", all_scores.shape)

전체 유사도 행렬 크기: (986, 986)


## 프롬프트 예시

```
recommend_books 함수로 986권 전체의 추천 품질을 점검하고 싶습니다.
모든 도서에 대해 Top 5를 구해서
1. 1위 추천의 유사도를 모으고
2. 미분류가 아닌 도서는 추천 5권 중 같은 분야 비율을 모아 주세요.
1위 유사도의 평균, 중앙값, 그리고 1위 유사도가 0인 책의 수와 비율을 출력해 주세요.
try/except 예외 처리는 넣지 마세요.
초보자가 읽기 쉽게 10줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [25]:
top1_scores = []
same_field_ratios = []

for i in range(len(df_reco)):
    recs = recommend_books(i, df_reco, tfidf_matrix, top_n=5)
    top1_scores.append(recs["similarity"].iloc[0])

    field = df_reco.loc[i, "분야"]
    if field != "미분류":
        same_field_ratios.append((recs["분야"] == field).mean())

top1_scores = pd.Series(top1_scores)

print("1위 추천 유사도 평균:", round(top1_scores.mean(), 3))
print("1위 추천 유사도 중앙값:", round(top1_scores.median(), 3))
print()
zero = int((top1_scores == 0).sum())
print(f"1위 추천조차 유사도 0인 책: {zero} / {len(df_reco)} ({zero / len(df_reco) * 100:.1f}%)")

1위 추천 유사도 평균: 0.333
1위 추천 유사도 중앙값: 0.299

1위 추천조차 유사도 0인 책: 261 / 986 (26.5%)


## 프롬프트 예시

```
1위 추천 유사도가 든 Series top1_scores가 있습니다.
0, 0~0.2, 0.2~0.5, 0.5~0.8, 0.8 이상 구간으로 나눠
구간별로 몇 권인지 보여 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [26]:
# 1위 추천 유사도를 구간별로 본다
bands = pd.cut(
    top1_scores,
    bins=[-0.01, 0, 0.2, 0.5, 0.8, 1.0],
    labels=["0 (겹침 없음)", "0~0.2", "0.2~0.5", "0.5~0.8", "0.8 이상"],
)
bands.value_counts().sort_index()

0 (겹침 없음)    261
0~0.2         54
0.2~0.5      433
0.5~0.8      164
0.8 이상        74
Name: count, dtype: int64

## 프롬프트 예시

```
추천 5권 중 같은 분야 비율 목록 same_field_ratios가 있습니다.
이 값이 무작위로 골랐을 때보다 얼마나 높은지 비교하고 싶습니다.
미분류를 뺀 분야 비율의 제곱합을 무작위 기대값으로 계산해서
평균 같은 분야 비율, 무작위 기대값, 몇 배인지 출력해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [27]:
# 추천 5권이 기준 도서와 같은 분야인 비율 vs 무작위로 골랐을 때 기대값
field_share = df_reco.loc[df_reco["분야"] != "미분류", "분야"].value_counts(normalize=True)
random_expect = (field_share ** 2).sum()

print(f"추천 5권 중 같은 분야 비율 평균: {np.mean(same_field_ratios):.3f}")
print(f"무작위로 5권을 골랐을 때 기대값 : {random_expect:.3f}")
print(f"무작위 대비                      : {np.mean(same_field_ratios) / random_expect:.1f}배")

추천 5권 중 같은 분야 비율 평균: 0.298
무작위로 5권을 골랐을 때 기대값 : 0.089
무작위 대비                      : 3.4배


### 결과 해석

**추천이 주제를 어느 정도 따라간다** —
추천 5권 중 같은 분야 비율은 **0.298**로, 무작위로 골랐을 때(0.089)보다 **약 3.4배** 높았다.
제목만 쓰는 단순한 방식인데도 분야와 관련 있는 책을 찾아낸다.

**하지만 4권 중 1권은 추천이 불가능하다** —
**261권(26.5%)** 은 1위 추천조차 유사도가 0이었다.
제목의 단어가 다른 어떤 책과도 겹치지 않는다는 뜻이다.
이 책들은 추천 결과 5권이 전부 의미 없는 책으로 채워진다.

**0.8 이상의 높은 유사도는 대부분 같은 책이나 같은 시리즈다** —
유사도가 매우 높으면 "비슷한 책"이라기보다 **같은 책의 다른 판본**일 가능성이 높다.

---
# 보충 3. 의미 없는 추천을 걸러낼 수 있을까

## 해야 할 일

블로그의 추천 함수는 **유사도가 0이어도 5권을 채운다.**
그리고 자기 자신과 같은 제목만 빼기 때문에, **추천 결과끼리 같은 제목이 겹치는 경우**도 남는다.

두 가지를 바꾼 개선 버전을 만들어 비교한다.

| 바꾼 점 | 이유 |
|---|---|
| 유사도가 0 이하면 추천하지 않는다 | 겹치는 단어가 없는 책은 추천 근거가 없다 |
| 추천 결과 안에서도 같은 제목은 한 번만 | 같은 책이 두 번 추천되는 것을 막는다 |

> 본 실습의 `recommend_books()`는 블로그 그대로 두고, 비교용으로 따로 만든다.

## 프롬프트 예시

```
recommend_books 함수로 986권 전체의 Top 5를 구했을 때
추천 5권 안에 같은 제목이 중복으로 들어간 책이 몇 권인지 세고 싶습니다.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [28]:
# 블로그 함수로는 추천 결과 안에 같은 제목이 겹치는 책이 몇 권인지 확인한다
overlap = 0
for i in range(len(df_reco)):
    recs = recommend_books(i, df_reco, tfidf_matrix, top_n=5)
    if recs["상품명"].nunique() < len(recs):
        overlap += 1

print("추천 5권 안에 같은 제목이 중복된 책:", overlap, "권")

추천 5권 안에 같은 제목이 중복된 책: 86 권


## 프롬프트 예시

```
recommend_books를 개선한 recommend_books_filtered 함수를 따로 만들고 싶습니다.
1. 유사도가 min_similarity(기본 0) 이하면 더 이상 추천하지 않고
2. 자기 자신과, 이미 추천한 제목과 같은 제목은 빼 주세요.
추천할 책이 없으면 빈 표를 돌려주세요.
5, 3, 8번 도서로 테스트해서 추천이 없으면 '추천할 수 있는 도서가 없습니다'라고 출력해 주세요.
try/except 예외 처리는 넣지 마세요.
초보자가 읽기 쉽게 10줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [29]:
def recommend_books_filtered(selected_index, df, tfidf_matrix, top_n=5, min_similarity=0.0):
    scores = cosine_similarity(tfidf_matrix[selected_index], tfidf_matrix).flatten()
    seen = {df.loc[selected_index, "상품명"]}   # 이미 나온 제목
    picked = []

    for idx in scores.argsort()[::-1]:
        if scores[idx] <= min_similarity:       # 겹치는 단어가 없으면 멈춘다
            break
        title = df.loc[idx, "상품명"]
        if idx == selected_index or title in seen:
            continue
        seen.add(title)
        picked.append(idx)
        if len(picked) >= top_n:
            break

    result = df.loc[picked, ["상품명", "출판사", "분야"]].copy()
    result["similarity"] = [round(float(scores[i]), 4) for i in picked]
    return result.reset_index(drop=True)

for idx in [5, 3, 8]:
    print(f"[{idx}] {df_reco.loc[idx, '상품명']}")
    recs = recommend_books_filtered(idx, df_reco, tfidf_matrix)
    if recs.empty:
        print("      추천할 수 있는 도서가 없습니다 (겹치는 단어 없음)")
    for _, row in recs.iterrows():
        print(f"      {row['similarity']:.3f}  {row['상품명'][:38]}")
    print()

[5] 테오
      추천할 수 있는 도서가 없습니다 (겹치는 단어 없음)

[3] 싯다르타
      0.363  싯다르타(초판본)(1922년 오리지널 초판본 표지디자인)

[8] 판매의 법칙
      0.430  불변의 법칙
      0.406  침투율의 법칙
      0.303  인간 본성의 법칙 (리커버:K)
      0.275  부자 아빠 투자 불변의 법칙



## 프롬프트 예시

```
개선 함수 recommend_books_filtered를 986권 전체에 적용해
추천 0권, 1~4권, 5권인 도서가 각각 몇 권인지 세고,
미분류가 아닌 도서의 같은 분야 비율을 블로그 함수 결과와 비교해 주세요.
try/except 예외 처리는 넣지 마세요.
초보자가 읽기 쉽게 10줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [30]:
# 전체 도서에 적용해서 비교한다
counts = []
same_field_v2 = []

for i in range(len(df_reco)):
    recs = recommend_books_filtered(i, df_reco, tfidf_matrix)
    counts.append(len(recs))
    field = df_reco.loc[i, "분야"]
    if field != "미분류" and len(recs) > 0:
        same_field_v2.append((recs["분야"] == field).mean())

counts = pd.Series(counts)
print("추천 0권 (추천 불가):", int((counts == 0).sum()))
print("추천 1~4권          :", int(((counts > 0) & (counts < 5)).sum()))
print("추천 5권            :", int((counts == 5).sum()))
print()
print(f"같은 분야 비율  블로그 함수 {np.mean(same_field_ratios):.3f}  ->  개선 함수 {np.mean(same_field_v2):.3f}")

추천 0권 (추천 불가): 261
추천 1~4권          : 295
추천 5권            : 430

같은 분야 비율  블로그 함수 0.298  ->  개선 함수 0.477


### 개선 결과

| 항목 | 블로그 함수 | 개선 함수 |
|---|---|---|
| 추천 5권 안에 같은 제목 중복 | 86권에서 발생 | 0 |
| 겹치는 단어 없는 책 추천 | 5권 억지로 채움 | **"추천 없음"으로 표시** (261권) |
| 추천 5권 중 같은 분야 비율 | 0.298 | **0.477** |

`테오`처럼 추천 근거가 없는 책에 화장지를 추천하는 대신 **"추천할 도서가 없다"** 고 솔직하게 알려준다.
그 결과 추천이 나온 책들의 같은 분야 비율이 **0.298 → 0.477**로 올랐다.

> 추천 개수를 억지로 채우는 것보다 **근거 없는 추천을 빼는 것**이 추천의 신뢰도를 높인다.
> 다만 26.5%의 책은 여전히 추천을 받지 못하므로, 제목 외의 정보(저자·분야·책 소개)가 필요하다.

---
# 실습 16. Chapter 06 Streamlit 연결 준비하기

## 해야 할 일

Chapter 06에서는 사용자가 직접 도서를 선택하도록 만든다.

```
st.selectbox() -> selected_index -> recommend_books() -> Top 5 -> st.dataframe()
```

선택 목록은 **index와 제목을 함께** 사용할 수 있다.
Chapter 06의 핵심은 새로운 추천 알고리즘을 만드는 것이 아니라,
**이번 Chapter에서 만든 추천 함수를 웹 UI에 연결하는 것**이다.

## 프롬프트 예시

```
Streamlit selectbox에 쓸 도서 선택 목록을 만들고 싶습니다.
df_reco의 각 행을 '번호 | 상품명' 문자열을 key로, index를 value로 하는
딕셔너리 book_options로 만들고, 개수와 앞 3개를 출력해 주세요.
함수로 만들지 말고 바로 실행되는 코드로 작성해 주세요.
try/except 예외 처리는 넣지 마세요.
5줄 이내로 짧게 작성해 주세요.
설명은 코드 옆 짧은 주석으로만 달아 주세요.
```

In [31]:
book_options = {
    f"{idx} | {row['상품명']}": idx
    for idx, row in df_reco.iterrows()
}

print("선택 목록 개수:", len(book_options))
print()
for label in list(book_options)[:3]:
    print(" ", label, "->", book_options[label])

선택 목록 개수: 986

  0 | 세네카, 오늘을 빼앗기고 있는 당신에게 -> 0
  1 | 흔한남매 23 -> 1
  2 | 머니 트렌드 2027 -> 2


같은 제목이 두 번 등록된 책도 **앞에 index가 붙어 있어서** 선택 목록에서 서로 구분된다.
제목만 key로 쓰면 뒤의 책이 앞의 책을 덮어써서 목록에서 사라진다.

---
# 실습 17. 전체 코드 흐름 정리하기

```
CSV 로드 -> 상품명 정리 -> index 재설정 -> TF-IDF fit_transform -> 기준 도서 선택
-> cosine similarity -> 유사도 정렬 -> 자기 자신 제외 -> Top 5 -> recommend_books() -> CSV 저장
```

> 이 코드는 흐름을 다시 확인하기 위한 정리용이다.
> 각 단계의 원리를 이해하지 않은 채 이 코드만 복사하는 것은 권장하지 않는다.

## 프롬프트 예시

```
이번 Chapter의 전체 흐름을 한 셀로 정리하고 싶습니다.
CSV 로드 -> 상품명 정리 -> index 재설정 -> TF-IDF -> 추천 함수
-> 0번 도서 추천 실행 -> CSV 저장 순서입니다.
단계마다 번호 주석을 달아 주세요.
try/except 예외 처리는 넣지 마세요.
```

In [32]:
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# 1. 데이터
df_reco = pd.read_csv(
    "book_bestseller_clean.csv",
    encoding="utf-8-sig",
)
df_reco["상품명"] = (
    df_reco["상품명"]
    .fillna("")
    .astype(str)
    .str.strip()
)
df_reco = (
    df_reco[df_reco["상품명"] != ""]
    .reset_index(drop=True)
)

# 2. TF-IDF
tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(
    df_reco["상품명"]
)

# 3. 추천 함수
def recommend_books(
    selected_index,
    df,
    tfidf_matrix,
    top_n=5,
):
    selected_title = df.loc[selected_index, "상품명"]
    scores = cosine_similarity(
        tfidf_matrix[selected_index],
        tfidf_matrix,
    ).flatten()
    sorted_indices = scores.argsort()[::-1]
    recommended_indices = []
    for idx in sorted_indices:
        if idx == selected_index:
            continue
        if df.loc[idx, "상품명"] == selected_title:
            continue
        recommended_indices.append(idx)
        if len(recommended_indices) >= top_n:
            break
    columns = [
        column
        for column in ["상품명", "인물", "출판사", "분야"]
        if column in df.columns
    ]
    result = df.loc[
        recommended_indices,
        columns,
    ].copy()
    result["similarity"] = [
        round(float(scores[idx]), 4)
        for idx in recommended_indices
    ]
    return result.reset_index(drop=True)

# 4. 추천 실행
selected_index = 0
recommendations = recommend_books(
    selected_index,
    df_reco,
    tfidf_matrix,
    top_n=5,
)

# 5. 저장
recommendations.to_csv(
    "chapter05_recommendations.csv",
    index=False,
    encoding="utf-8-sig",
)

recommendations

,상품명,출판사,분야,similarity
0,"오늘을 빼앗기는 당신에게, 세네카",유페이퍼,미분류,0.6682
1,돌이킬 수 있는,아작,소설,0.2521
2,남아 있는 나날,민음사,소설,0.2039
3,비전공자도 이해할 수 있는 LLM 수업,비즈니스북스,과학,0.1617
4,품격 있는 대화를 위한 지식 브리핑,북플레저,인문,0.1475


---
# 결과 정리

> 아래 숫자는 위 셀들의 **실제 실행 결과**를 보고 적은 것이다.

## Chapter 05 결과

### 데이터
- 추천 대상: `book_bestseller_clean.csv`의 상품명 **986권** (빈 제목 없음)
- 사용 컬럼: 상품명(추천 근거), 출판사·분야(결과 검토용)
- `인물` 컬럼은 Chapter 01에서 `저자`로 바꿔 저장해 결과에 나오지 않는다

### 방법
- TfidfVectorizer(기본 설정) -> **986 x 2,179** 행렬
- 코사인 유사도로 선택 도서와 전체 도서 비교
- 자기 자신과 동일 제목을 제외하고 Top 5

### 예시 결과 — `세네카, 오늘을 빼앗기고 있는 당신에게`

| 순위 | 추천 도서 | 분야 | 유사도 |
|---|---|---|---|
| 1 | 오늘을 빼앗기는 당신에게, 세네카 | 미분류 | 0.6682 |
| 2 | 돌이킬 수 있는 | 소설 | 0.2521 |
| 3 | 남아 있는 나날 | 소설 | 0.2039 |
| 4 | 비전공자도 이해할 수 있는 LLM 수업 | 과학 | 0.1617 |
| 5 | 품격 있는 대화를 위한 지식 브리핑 | 인문 | 0.1475 |

1위는 같은 책의 다른 판본, 2~5위는 `있는`이라는 흔한 단어 하나만 겹친 결과였다.

### 전체 도서 기준 품질
- 1위 추천 유사도: 평균 **0.333**, 중앙값 **0.299**
- **1위 추천조차 유사도 0인 책: 261권 (26.5%)** -> 추천 근거가 없음
- 추천 5권 중 같은 분야 비율: **0.298** (무작위 0.089의 약 3.4배)
- 추천 결과 안에 같은 제목이 중복된 책: **86권**

### 개선 실험 (비교용)
- 유사도 0 이하 제외 + 추천 결과 안 중복 제목 제외
- 같은 분야 비율 **0.298 -> 0.477**, 대신 261권은 "추천 없음"

---
# 핵심 시사점

### 1. 제목 기반 추천은 "시리즈 찾기"에는 강하다
`흔한남매 23`의 추천 5권이 모두 흔한남매 시리즈였다.
제목에 고유한 시리즈명이나 브랜드명이 있으면 콘텐츠 기반 추천이 잘 작동한다.

### 2. 하지만 흔한 단어 하나로도 추천이 된다
세네카 책의 2~5위는 `있는`이라는 단어 하나만 겹친 책이었다.
`머니 트렌드 2027`에는 `2027`만 겹치는 수험서가 나왔다.
**유사도가 0보다 크다는 것이 "주제가 비슷하다"는 뜻은 아니다.**

### 3. 4권 중 1권은 추천할 근거가 없다
986권 중 **261권(26.5%)** 은 제목 단어가 다른 어떤 책과도 겹치지 않았다.
그런데도 블로그 함수는 5권을 채워서 `테오`에 `화장지 30롤`을 추천했다.
**추천 개수를 채우는 것보다 근거 없는 추천을 빼는 것**이 더 중요하다.
개선 함수로 걸러내자 같은 분야 비율이 0.298에서 0.477로 올랐다.

### 4. 데이터 문제가 추천에도 그대로 나타난다
같은 책이 `미분류`와 실제 분야로 **두 번 등록**되어 있어 추천에 같은 제목이 중복되었다(86권).
화장지, 북커버 키트 같은 **비도서 상품**이 추천 결과에 섞였다.
Chapter 01에서 발견한 데이터 문제가 Chapter 05까지 이어진다.

### 5. 높은 유사도가 좋은 추천은 아니다
1위 추천 유사도가 0.8 이상인 74권을 보면 대부분 **같은 시리즈의 다른 권, 세트, 다른 판본**이었다.

```
1.000  인간의 굴레에서 2      <->  인간의 굴레에서 1
0.825  눈물을 마시는 새 세트  <->  [세트] 눈물을 마시는 새 (전 4권)
0.803  해커스 토익 700+       <->  해커스 토익 900+
```

`인간의 굴레에서 1`과 `2`가 유사도 **1.0**인 이유는, 기본 토큰 규칙이 **한 글자짜리 `1`, `2`를 버려서**
두 제목이 완전히 같은 벡터가 되기 때문이다.
사용자에게는 이미 고른 책과 거의 같은 책을 추천하는 셈이라 오히려 쓸모가 적을 수 있다.

### 6. 유사도는 취향 점수가 아니다
이번 추천은 구매 이력·평점·판매량을 전혀 쓰지 않았다.
"현재 TF-IDF 제목 표현 기준으로 유사도가 높은 도서"라고만 말할 수 있다.

## 개선 방향

| 방향 | 기대 효과 | 근거 |
|---|---|---|
| 유사도 하한선을 두고 근거 없으면 추천하지 않기 | 무의미한 추천 제거 | 유사도 0인 책 261권 |
| 추천 결과 안에서도 같은 제목 중복 제거 | 같은 책 반복 추천 방지 | 중복 86권 |
| 비도서 상품과 중복 등록 데이터 정리 | 화장지·굿즈 추천 방지 | 보충 1 `테오` 사례 |
| 연도 숫자와 흔한 단어를 불용어로 | 엉뚱한 공통점 제거 | `2027`, `있는` 사례 |
| Kiwi로 명사만 추출해서 벡터화 | 조사 때문에 못 찾는 단어 해결 | Chapter 04 오분류 원인 |
| 저자·분야·책 소개를 함께 사용 | 제목만으로 부족한 정보 보완 | 추천 불가 26.5% |

---
# 최종 체크리스트

- [x] 전처리 데이터를 불러왔다.
- [x] 상품명을 정리하고 index를 재설정했다.
- [x] 도서 제목을 TF-IDF 벡터로 변환했다.
- [x] 코사인 유사도의 의미를 설명할 수 있다.
- [x] 선택 도서와 전체 도서의 유사도를 계산했다.
- [x] 유사도를 높은 순서로 정렬했다.
- [x] 자기 자신을 제외했다.
- [x] Top 5 추천 결과를 만들었다.
- [x] 추천 함수를 만들었다.
- [x] 추천 결과를 실제 제목과 비교했다.
- [x] 결과를 CSV로 저장했다.
- [x] 제목 기반 추천의 한계를 설명할 수 있다.
- [x] Chapter 06 Streamlit 연결 구조를 이해했다.

---
# 이번 Chapter에서 꼭 기억할 5가지

1. TF-IDF 벡터는 분류뿐 아니라 **추천에도** 사용할 수 있다.
2. 코사인 유사도는 두 벡터의 **방향이 얼마나 비슷한지** 비교한다.
3. 자기 자신과의 유사도가 가장 높으므로 추천에서 **제외**한다.
4. **유사도 내림차순 -> 자기 자신 제외 -> Top N**이 기본 추천 흐름이다.
5. 제목 유사도는 사용자의 취향이나 책의 품질을 의미하지 않는다.

---
# 다음 Chapter 연결

```
Chapter 04   새 도서 제목 -> TF-IDF -> Naive Bayes -> 예상 분야
Chapter 05   기존 도서 선택 -> TF-IDF -> Cosine Similarity -> 유사 도서 Top 5
```

Chapter 06에서는 이 두 기능을 Streamlit으로 연결한다.

```
사용자
   ↓
Streamlit UI
   ├─ 제목 입력 -> 분야 분류
   └─ 도서 선택 -> Top 5 추천
```

## Chapter 05 한 문장 정리

> 도서 제목을 TF-IDF 벡터로 표현하고, 선택한 도서와 전체 도서의 코사인 유사도를 계산한 뒤
> 자기 자신을 제외하고 유사도가 높은 Top 5를 선택하면 제목 기반 콘텐츠 추천을 만들 수 있다.